### PR TITLE
First pass at building out FeeRouter

### PR DESCRIPTION
**Pull request by alexanderattar**
# Build FeeRouter: Implement Cross-chain Fee Management System (Base -> Ethereum)

This PR implements the complete fee management system for Holograph Doppler Launchpad. The system collects protocol fees on Base, bridges them to Ethereum, and distributes HLG rewards to stakers.

### New Omnichain FeeRouter
- Built FeeRouter supporting Base ↔ Ethereum operations
- LayerZero V2 integration for cross-chain fee bridging
- Uniswap V3 integration for WETH→HLG swaps (0.3% fee tier)
- 50/50 burn/stake distribution logic
- Proper LZ options encoding 

### Security Enhancements
- Added trusted remote validation for LayerZero messages
- Implemented pausable functionality for emergency stops
- Added protocol fee percentage configuration (1.5% of launch fees)
- Slippage protection for DEX swaps

### Contract Updates
- Renamed `notifyReward` → `addRewards` in StakingRewards per spec
- Updated HolographFactory to forward only protocol fee portion
- Added admin functions for fee management

### Interface Cleanup
- Extracted all inline interfaces to dedicated files in `src/interfaces/`
- Fixed compilation issues and state mutability conflicts
- Updated IFeeRouter with omnichain methods

### Testing Infrastructure
- Integration test suite
- Mock contracts for LayerZero, WETH, and Uniswap components
- End-to-end fee flow testing from Base to Ethereum
- Security validation and edge case coverage

### Documentation
- Added README with system architecture diagrams
- Security features and testing documentation

## Technical Details

**Fee Flow**: Token launches on Base generate 1.5% protocol fees → FeeRouter bridges ETH to Ethereum → Swap to HLG → 50% burned, 50% distributed to stakers

**Cross-Chain**: Uses LayerZero V2 with proper trusted remote validation and nonce-based replay protection

**Testing**: All tests passing (16/16) with coverage of core functionality, security, and edge cases

image

### CODE REVIEW

File: src/HolographFactory.sol [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    /* -------------------------------------------------------------------------- */
+    /*                                    OWNER                                   */
+    /* -------------------------------------------------------------------------- */
+    function setLaunchFee(uint256 weiAmount) external onlyOwner {
+        launchFeeETH = weiAmount;
+    }
+
+    function setProtocolFeePercentage(uint256 newPercentage) external onlyOwner {
```
Comment (sogoiii) [inline src/HolographFactory.sol:143]: @alexanderattar remind me why we have this here? were we gunna take a fee on every bridge as well?

File: src/HolographFactory.sol [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        if (msg.value > launchFeeETH) payable(msg.sender).transfer(msg.value - launchFeeETH);
+
+        (asset, , , , ) = dopplerAirlock.create(params);
+        emit TokenLaunched(asset, params.salt);
+    }
+
+    /* -------------------------------------------------------------------------- */
+    /*                               CROSS-CHAIN SEND                             */
+    /* -------------------------------------------------------------------------- */
+    function bridgeMint(
```
[the remark this answers is not in the document]
Comment (alexanderattar) [reply src/HolographFactory.sol:104]: yeah that's true. i called it that since the payload is a `mint` call on the destination chain. i can just call it `bridge` or `bridgeOut`, `bridgeToken`. Lmk what you think

File: src/HolographFactory.sol (hunk 1 above)
Comment (alexanderattar) [reply src/HolographFactory.sol:170]: This sets the protocol fee (currently 1.5% in basis points) as was defined in the original struct. This is only being taken at createToken time at the moment, but I'm glad you brought that up because we're supposed to be taking it for bridge txns too.. adding that to my notes of things to update

File: src/StakingRewards.sol [hunk 3]
```diff
@@ -163,67 +163,66 @@ contract StakingRewards is Ownable, ReentrancyGuard, Pausable {
...
-     * @notice Called by the FeeRouter when HLG protocol fees are sent here.
-     * @dev    Distributes the reward immediately by updating
-     *         `rewardPerTokenStored`.
-     * @param  amount Amount of HLG transferred from the FeeRouter.
+     * @notice Called by the FeeRouter when HLG protocol fees are sent here
+     * @dev    Distributes the reward immediately by updating rewardPerTokenStored
+     * @param  amount Amount of HLG transferred from the FeeRouter
      */
-    function notifyReward(uint256 amount) external nonReentrant updateReward(address(0)) {
+    function addRewards(uint256 amount) external nonReentrant updateReward(address(0)) {
```
[the remark this answers is not in the document]
Comment (alexanderattar) [reply src/StakingRewards.sol:181]: yeah, it's for security. without addRewards() the FeeRouter could send tokens but they'd just sit in the contract balance without updating rewardPerTokenStored, so users couldn't actually claim them. it also makes sure only the feerouter can add the rewards via `FeeRouterOnly` and also emits an event
